pkg/metaclient/namespace/kv.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//[reference]: https://github.com/etcd-io/etcd/blob/aa75fd08509db3aea8939cdad44e1ee9b8157b8c/client/v3/namespace/kv.go
+
+package namespace
+
+import (
+	"context"
+
+	cerrors "github.com/hanfei1991/microcosm/pkg/errors"
+	"github.com/hanfei1991/microcosm/pkg/metaclient"
+	"github.com/hanfei1991/microcosm/pkg/metaclient/internal"
+)
+
+type kvPrefix struct {
+	internal.KVEx
+	pfx string
+}
+
+// NewPrefixKV wraps a KVEx instance so that all requests
+// are prefixed with a given string.
+func NewPrefixKV(kv internal.KVEx, prefix string) metaclient.KV {
+	return &kvPrefix{kv, prefix}
+}
+
+func (kv *kvPrefix) Put(ctx context.Context, key, val string) (*metaclient.PutResponse, error) {
+	if len(key) == 0 {
+		return nil, cerrors.ErrMetaEmptyKey
+	}
+	op := kv.prefixOp(metaclient.OpPut(key, val))
+	r, err := kv.KVEx.Do(ctx, op)
+	if err != nil {
+		return nil, err
+	}
+	put := r.Put()
+	return put, nil
+}
+
+func (kv *kvPrefix) Get(ctx context.Context, key string, opts ...metaclient.OpOption) (*metaclient.GetResponse, error) {
+	if len(key) == 0 && !(metaclient.IsOptsWithFromKey(opts) || metaclient.IsOptsWithPrefix(opts)) {
+		return nil, cerrors.ErrMetaEmptyKey
+	}
+	r, err := kv.KVEx.Do(ctx, kv.prefixOp(metaclient.OpGet(key, opts...)))
+	if err != nil {
+		return nil, err
+	}
+	get := r.Get()
+	kv.unprefixGetResponse(get)
```

Comment:
Supposing the namespace has prefix `/prefix1`, when user queries with `WithFromKey` option, and no key starting from `/prefix1` is matched, and returns a key with prefix `/prefix2`, then user accesses invalid data. 

such as querying `/zz` with `WithFromKey` option. It will return `/prefix2/aa`

```
/prefix1/aa
/prefix1/bb
/prefix2/aa
```

Reply:
It will not happen.
Actually,  as for your case, internal key span will be [/prefix1, /prefix2)

Reply:
Which code generates the `/prefix2` part in `[/prefix1, /prefix2)`

Reply:
prefixOp()--->prefixInterval()